src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditWaiter.cs

```diff
@@ -10,21 +9,65 @@
+        // _registration are zero'd out upon completion, because they're used for synchronization
+        // between successful completion and cancellation.  This means an instance may end up
+        // referencing the underlying CancellationTokenSource even after the await operation has completed.
+
+        /// <summary>Cancellation token for the current wait operation.</summary>
+        private CancellationToken _cancellationToken;
+        /// <summary>Cancellation registration for the current wait operation.</summary>
+        private CancellationTokenRegistration _registration;
+        /// <summary><see cref="IValueTaskSource"/> implementation.</summary>
+        private ManualResetValueTaskSourceCore<int> _source;
+
+        // State carried with the waiter for the consumer to use; these aren't used at all in the implementation.
+
+        /// <summary>Amount of credit desired by this waiter.</summary>
         public int Amount;
+        /// <summary>Next waiter in a list of waiters.</summary>
         public CreditWaiter? Next;
-        protected ManualResetValueTaskSourceCore<int> _source;
 
-        public CreditWaiter() => _source.RunContinuationsAsynchronously = true;
+        /// <summary>Initializes a waiter for a credit wait operation.</summary>
+        /// <param name="cancellationToken">The cancellation token for this wait operation.</param>
+        public CreditWaiter(CancellationToken cancellationToken)
+        {
+            _source.RunContinuationsAsynchronously = true;
+            RegisterCancellation(cancellationToken);
+        }
 
-        public ValueTask<int> AsValueTask() => new ValueTask<int>(this, _source.Version);
+        /// <summary>Re-initializes a waiter for a credit wait operation.</summary>
+        /// <param name="cancellationToken">The cancellation token for this wait operation.</param>
+        public void ResetForAwait(CancellationToken cancellationToken)
+        {
+            _source.Reset();
+            RegisterCancellation(cancellationToken);
+        }
 
-        public bool IsPending => _source.GetStatus(_source.Version) == ValueTaskSourceStatus.Pending;
+        /// <summary>Registers with the cancellation token to transition the source to a canceled state.</summary>
+        /// <param name="cancellationToken">The cancellation token with which to register.</param>
+        private void RegisterCancellation(CancellationToken cancellationToken)
+        {
+            _cancellationToken = cancellationToken;
+            _registration = cancellationToken.UnsafeRegister(static s =>
+            {
+                // The callback will only fire if cancellation owns the right to complete the instance.
+                var thisRef = (CreditWaiter)s!;
+                thisRef._source.SetException(ExceptionDispatchInfo.SetCurrentStackTrace(new OperationCanceledException(thisRef._cancellationToken)));
```

Comment:
I believe even it had one, simply reporting a cancellation wouldn't be enough because we also capture the stack trace here.

Reply:
I'm never quite sure when we want to capture the stack trace in cases like this, and when we don't. But that's unrelateed to this PR.

Reply:
> This is necessary because MVRTSC doesn't have a SetCanceled method?

Yes. It also doesn't really matter here because it's always awaited, so the consumer didn't see any difference.

> I'm never quite sure when we want to capture the stack trace in cases like this, and when we don't.

When I initially rolled it out, I added it in places where a) I thought the additional data would be helpful for diagnostics and b) the overhead wouldn't be impactful.